torch_geometric/nn/models/basic_pools.py

```diff
@@ -0,0 +1,316 @@
+    r"""An abstract class for implementing basic graph pooling architectures.
+
+    Args:
+        in_channels (int): Size of each input sample.
+        hidden_channels (int): Size of each hidden and output sample.
+        num_layers (int): Number of message passing layers.
+        dropout (float, optional): Dropout probability. (default: :obj:`0.`)
+        act (Callable, optional): The non-linear activation function to use.
+            (default: :meth:`torch.nn.ReLU(inplace=True)`)
+        norm (torch.nn.Module, optional): The normalization operator to use.
+            (default: :obj:`None`)
+        jk (str, optional): The Jumping Knowledge mode
+            (:obj:`"last"`, :obj:`"cat"`, :obj:`"max"`, :obj:`"last"`).
+            (default: :obj:`"last"`)
+        glob_pool (str, optional): The global pooling operator to use
+            (:obj:`"add"`, :obj:`"max"`, :obj:`"mean"`).
+            (default: :obj:`"mean"`)
+    """
+    def __init__(self, in_channels: int, hidden_channels: int, num_layers: int,
+                 dropout: float = 0.0,
+                 act: Optional[Callable] = ReLU(inplace=True),
+                 norm: Optional[torch.nn.Module] = None, jk: str = 'last',
+                 glob_pool: str = 'mean'):
+        super().__init__()
+        self.in_channels = in_channels
+        self.hidden_channels = hidden_channels
+        self.out_channels = hidden_channels
+        if jk == 'cat':
+            self.out_channels = num_layers * hidden_channels
+        self.num_layers = num_layers
+        self.dropout = dropout
+        self.act = act
+
+        self.convs = ModuleList()
+        self.pools = ModuleList()
+
+        self.jk = None
+        if jk != 'last':
+            self.jk = JumpingKnowledge(jk, hidden_channels, num_layers)
+
+        self.norms = None
+        if norm is not None:
+            self.norms = ModuleList(
+                [copy.deepcopy(norm) for _ in range(num_layers)])
+
+        if glob_pool == 'mean':
+            self.global_pool = global_mean_pool
+        elif glob_pool == 'max':
+            self.global_pool = global_max_pool
+        elif glob_pool == 'add':
+            self.global_pool = global_add_pool
+
+    def reset_parameters(self):
+        for conv in self.convs:
+            conv.reset_parameters()
+        for norm in self.norms or []:
+            norm.reset_parameters()
+        if self.jk is not None:
+            self.jk.reset_parameters()
+
+    def forward(self, x: Tensor, edge_index: Adj, batch: Tensor, *args, **kwargs) -> Tensor:
+        xs: List[Tensor] = []
+        for i in range(self.num_layers):
+            x = self.convs[i](x, edge_index, *args, **kwargs)
+            if self.norms is not None:
+                x = self.norms[i](x)
+            if self.act is not None:
+                x = self.act(x) 
+
+            xs += [self.global_pool(x, batch)]
+            if i % 2 == 0 and i < len(self.convs) - 1:
+                pool = self.pools[i // 2]
```

Comment:
I think we should create a function `self.pool(layer, x, edge_index, batch, *kwargs)` that takes care of pooling, and that one needs to override. I think not all pooling ops are perfectly aligned to this interface yet.